sfge/src/main/java/com/salesforce/rules/PerformNullCheckOnSoqlVariables.java

```diff
@@ -0,0 +1,149 @@
+    protected String getUrl() {
+        return URL;
+    }
+
+    public static PerformNullCheckOnSoqlVariables getInstance() {
+        return LazyHolder.INSTANCE;
+    }
+
+    @Override
+    protected boolean isEnabled() {
+        return true;
+    }
+
+    @Override
+    public boolean test(BaseSFVertex vertex) {
+        return false;
+    }
+
+    /**
+     * Note: this rule runs differently. Since the {@link SymbolProvider} is needed to determine
+     * whether a value might be null (and the symbol provider is unavailable during path walking and
+     * therefore the _run method), this rule only accepts vertices that are confirmed to be a
+     * violation. This is because the rule uses {@link Constraint}s to see if something has been
+     * null-checked, which are exclusively available during path expansion. Therefore, when walking
+     * the path, the rule automatically knows any accepted vertex is a violation.
+     */
+    @Override
+    public boolean test(BaseSFVertex vertex, SymbolProvider symbols) {
+
+        // We only check for SOQL expressions with variables as grandchildren.
+        // No need to check for MethodCallExpressionVertex because the only relevant method is
+        // Database.queryWithBinds, though that method already does null checks for the keys and
+        // values in its bind variables.
+        if (!(vertex instanceof VariableExpressionVertex)
+                || !(vertex.getParent().getParent() instanceof SoqlExpressionVertex)) {
+            return false;
+        }
+
+        VariableExpressionVertex methodVertex = (VariableExpressionVertex) vertex;
+
+        // try and find the value of this vertex
+        Optional<ApexValue<?>> variableValueOpt =
+                ScopeUtil.resolveToApexValue(symbols, methodVertex);
+        if (variableValueOpt.isPresent()) {
+            ApexValue<?> variableValue = variableValueOpt.get();
+
+            if (variableValue.hasNegativeConstraint(Constraint.Null)) {
+                // negative constraint means the value is guaranteed NOT to be null,
+                // so this deserves no violation
+                return false;
+            } else if (variableValue.isNull() || variableValue.isIndeterminant()) {
+                // if the variable is null (is actually null or has positive null constraint) OR the
+                // value is indeterminant, this deserves a violation
+                return true;
+            }
+
+        } else {
+            throw new UnexpectedException(
+                    "PerformNullCheckOnSoqlVariables couldn't find an apex value associated with variable vertex "
+                            + vertex);
+        }
+        return false;
+    }
+
+    /**
+     * Note: this rule runs differently. The {@link #test(BaseSFVertex, SymbolProvider)} method
+     * above only accepts vertices that are confirmed to be violations. So, this run method just
+     * needs to actually create and return the violation.
+     */
```

Comment:
It's important to have a comment conveying this information, but a JDoc header might be the wrong place for it. You could probably just do a comment somewhere within the method itself.